sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/plans/logical/cteOperators.scala

```diff
@@ -150,6 +150,22 @@ case class CTERelationDef(
       false
     }
   }
+  lazy val hasSelfReferenceInSubCTE: Boolean = {
+    val withCTENode: Option[WithCTE] = child match {
+      case SubqueryAlias(_, withCTE @ WithCTE(_, _)) =>
+        Some(withCTE)
+      case SubqueryAlias(_, UnresolvedSubqueryColumnAliases(_, withCTE @ WithCTE(_, _))) =>
+        Some(withCTE)
+      case _ => None
+    }
+    if (withCTENode.isDefined) {
+      withCTENode.get.cteDefs.map(_.collectFirstWithSubqueries {
```

Comment:
Maybe we could stop at the first occurrance with:
```
!withCTENode.exists(!_.cteDefs.exists(_.collectFirstWithSubqueries {
  case CTERelationRef(this.id, _, _, _, _, true, _) => true
}.isDefined))
```

Reply:
+1